test/small_granularity.c

```diff
@@ -0,0 +1,117 @@
+#include <small/small.h>
+#include <small/quota.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include "unit.h"
+
+enum {
+	GRANULARITY_MIN = 4,
+	GRANULARITY_MAX = 1024,
+	ALLOCATION_COUNT = 10,
+};
+
+struct slab_arena arena;
+struct slab_cache cache;
+struct small_alloc alloc;
+struct quota quota;
+
+static unsigned *ptrs[FACTOR_POOL_MAX];
+
+static inline void
+free_checked(struct mempool *pool, unsigned *ptr)
+{
+	/*
+	 * Checking previous saved data.
+	 */
+	fail_unless(ptr[0] < FACTOR_POOL_MAX &&
+		    ptr[pool->objsize/sizeof(unsigned)-1] == ptr[0]);
+	int pos = ptr[0];
+	fail_unless(ptrs[pos] == ptr);
+	ptrs[pos][0] = ptrs[pos][pool->objsize/sizeof(unsigned)-1] = INT_MAX;
+	mempool_free(pool, ptrs[pos]);
+	ptrs[pos] = NULL;
+}
+
+static inline void *
+alloc_checked(struct mempool *pool, unsigned pos)
+{
+	if (ptrs[pos]) {
```

Comment:
I don't get how ptrs[pos] can be non-null: before starting second test (invoking alloc/free_checked) array of ptrs is nullified..

Reply:
You're right! change to fail_unless(ptrs[pos] == NULL)